### ISSUE TITLE
Suggestion to persist Article along with its Categories

### ISSUE THREAD

**Issue by dldharma** [infolab-csail/WikipediaBase#277]
Every Article has 1 or more Categories (persisted in or derived from Article.markup)

## Example: [Tom Cruise](https://en.wikipedia.org/wiki/Tom_Cruise)
Some of the Categories to which the Article belongs to are: 
[American film producers](https://en.wikipedia.org/wiki/Category:American_film_producers)
[American male film actors](https://en.w
<img width="547" alt="screen shot 2016-10-08 at 11 06 44 am" src="https://cloud.githubusercontent.com/assets/9015969/19210732/3003fff4-8d49-11e6-9988-70b94b033d10.png">
<img width="633" alt="screen shot 2016-10-08 at 11 07 50 am" src="https://cloud.githubusercontent.com/assets/9015969/19210734/3be245c4-8d49-11e6-9cfe-2361027676c6.png">

ikipedia.org/wiki/Category:American_male_film_actors)
Each of these Categories in turn may have Parent as well Child Categories (PSA screenshots).

## Analysis (after successful setup following [readme](https://github.com/infolab-csail/WikipediaBase)) (PSA screenshots):
We looked at the [Article.py](https://github.com/infolab-csail/WikipediaBase/blob/master/wikipediabase/article.py) model
There is a method categories() which returns a list of string. PSA code below.
In turn, these categories are derived from Article.markup.
**It is these categories which we are primarily interested in as it can be leveraged for Classification and Discovery.**
Every Article contains these categories. 
Infact, we found that every Articles may NOT contain association with Classes.

## Suggestion:
While persisting each Article, we may also persist it's relevant classes. Something like:
- Master: Flat list of categories // category_id , category_name
- One to Many: Article Categories // article_id , category_id
- Many to Many: Category relations or mappings // parent_category_id , child_category_id

**Comment by chsamlee** [comment]
I agree that categories are useful for classification and discovery, but I have 2 concerns:

1. Backend generation. Right now it takes 5 days to generate a new backend, and ideally the process should not take more than 7 days due to some CSAIL restrictions. If we're adding the categories to DB while generating the backend, we need to make sure that the total generation time doesn't go over the limit. However, I think Wikipedia has a category hierarchy dump so maybe we can use that instead of generating the hierarchy ourselves (and maybe we can even do it right after loading articles to DB).

2. Speed of category-related queries. Because of the frequency of `get-classes` calls, classification functions must run fast. We cannot traverse it in the same way we traverse through redirects (analyze the object, query the DB for next object, repeat) because the category hierarchy graph is deep and DB calls takes a long time to run (~30 ms per query). I can think of two potential solutions to this problem:
   - For category hierarchy only, ditch RDBMS and use a graph database instead. Then most work will be on integrating this feature with the rest of the system that runs on Postgres.
   - Split the category graph into components and fetch 1 component at a time (instead of 1 item). The main work will be on finding a good way to split the graph.

**Comment by dldharma** [comment]
Second your opinion. 

Possible approach (just a suggestion): **Consider these as extensions**.
Separate or isolated scripts - which after DATA IMPORT - perform optionally required action(s). Thus, whoever needs it, will have the liberty to execute them.

**What we are already working on is**:
1. A Python DB script which works exactly as commented above. 
    For each Article, persist categories to Master List as well as a Category Article Mapping table.
2. Enhance the existing API library to get categories. 
    Like "get-classes" we can provide "get-categories" implementing in KnowledgeBase class etc. 
3. Identifying sub categories. Another script which will pick each category in Master List, invoke WIKI API to fetch sub categories. Persist category parent and child in a category relation table.

We would then look at specific level / depth of categories - in association with classes - and figure out - what do we classify this as ? This is our todo and will require some MANUAL work to TRAIN / ANALYSE and tweak this strategy accordingly.

We have already tested neo4j GraphDB and our exploring cipher to traverse multi level node relations.  The big vision is to ideally convert this whole RDBMS model to Graph.

**Comment by alvaromorales** [comment]
Great suggestions!

If we pursue this, we should definitely extend KnowledgeBase to have a `get-categories` method instead of incorporating categories into `get-classes`. Classes and categories are not the same thing, so we shouldn't conflate them.

The WikipediaBase data model is pretty tightly coupled to an RDBMS. I would be very hesitant to add a dependency to another database system besides Postgres. This would make the process of deploying WikipediaBase more complex.

Adding categories as an optional step to the backend generation process is a good idea. We could add a command-line option to the `build_backend.py` script that skips categories by default, unless the user specifically enables the flag.

Wikipedia has SQL dumps for category data. See `enwiki-latest-category.sql.gz` and `enwiki-latest-categorylinks.sql.gz` in the dumps [page](https://dumps.wikimedia.org/enwiki/latest/). See also this StackOverflow [answer](http://stackoverflow.com/questions/17432254/wikipedia-category-hierarchy-from-dumps).

**It seems like a lot of people have worked on this problem before**, so we should do research before diving into a solution.

In terms of prioritization, this enhancement doesn't seem that high priority right now—we haven't yet figured out how to make good use of `get-types`, for example.

**Comment by dldharma** [comment]
Sure thing.
Meanwhile, we will proceed with needful to meet our milestones.
